### PR TITLE
define mission protocol

### LINKED ISSUE #5: Create proto file for mission plugin

### PR DESCRIPTION
**Pull request by JonasVautherin**
Fixes #5.

### CODE REVIEW

File: mission/mission.proto [hunk 1]
```diff
@@ -6,25 +6,63 @@ option java_package = "io.dronecore.mission";
...
-    rpc SendMission(SendMissionRequest) returns(MissionResult) {}
-    rpc StartMission(StartMissionRequest) returns(MissionResult) {}
+    rpc UploadMission(UploadMissionRequest) returns(UploadMissionResponse) {}
+    rpc DownloadMission(DownloadMissionRequest) returns(DownloadMissionResponse) {}
+    rpc StartMission(StartMissionRequest) returns(StartMissionResponse) {}
+    rpc PauseMission(PauseMissionRequest) returns(PauseMissionResponse) {}
+    rpc SetCurrentMissionItemId(SetCurrentMissionItemIdRequest) returns(SetCurrentMissionItemIdResponse) {}
+    rpc GetCurrentMissionItemId(GetCurrentMissionItemIdRequest) returns(GetCurrentMissionItemIdResponse) {}
+    rpc GetMissionSize(GetMissionSizeRequest) returns(GetMissionSizeResponse) {}
+    rpc SubscribeMissionProgress(SubscribeMissionProgressRequest) returns(stream MissionProgressResponse) {}
```
Comment (Rjasuja) [inline mission/mission.proto]: We can add ClearMission to delete all mission items at once

Comment (julianoes) [inline mission/mission.proto]: We now have sync and async functionality about the current mission.
@JonasVautherin do you think we should expose both through gRPC or only one of the two. This applies generall, I guess.

Comment (shakthi-prashanth-m) [reply mission/mission.proto]: It should be supported from DroneCore C++ first :)

Comment (Rjasuja) [reply mission/mission.proto]: It is part of Mavlink message specs

File: mission/mission.proto [hunk 2]
```diff
@@ -6,25 +6,63 @@ option java_package = "io.dronecore.mission";
...
 service MissionService {
-    rpc SendMission(SendMissionRequest) returns(MissionResult) {}
-    rpc StartMission(StartMissionRequest) returns(MissionResult) {}
+    rpc UploadMission(UploadMissionRequest) returns(UploadMissionResponse) {}
+    rpc DownloadMission(DownloadMissionRequest) returns(DownloadMissionResponse) {}
+    rpc StartMission(StartMissionRequest) returns(StartMissionResponse) {}
+    rpc PauseMission(PauseMissionRequest) returns(PauseMissionResponse) {}
+    rpc SetCurrentMissionItemId(SetCurrentMissionItemIdRequest) returns(SetCurrentMissionItemIdResponse) {}
+    rpc GetCurrentMissionItemId(GetCurrentMissionItemIdRequest) returns(GetCurrentMissionItemIdResponse) {}
+    rpc GetMissionSize(GetMissionSizeRequest) returns(GetMissionSizeResponse) {}
```
Comment (shakthi-prashanth-m) [inline mission/mission.proto]:
```
message GetMissionCountRequest {}
message GetMissionCountResponse {
    int32 mission_count = 1;
}
rpc GetMissionCount(GetMissionCountRequest) returns(GetMissionCountResponse) {}
```
Isn't this intuitive ?

File: mission/mission.proto [hunk 3]
```diff
@@ -6,25 +6,63 @@ option java_package = "io.dronecore.mission";
 option java_outer_classname = "MissionProto";
 
 service MissionService {
-    rpc SendMission(SendMissionRequest) returns(MissionResult) {}
-    rpc StartMission(StartMissionRequest) returns(MissionResult) {}
+    rpc UploadMission(UploadMissionRequest) returns(UploadMissionResponse) {}
+    rpc DownloadMission(DownloadMissionRequest) returns(DownloadMissionResponse) {}
+    rpc StartMission(StartMissionRequest) returns(StartMissionResponse) {}
+    rpc PauseMission(PauseMissionRequest) returns(PauseMissionResponse) {}
+    rpc SetCurrentMissionItemId(SetCurrentMissionItemIdRequest) returns(SetCurrentMissionItemIdResponse) {}
```
Comment (shakthi-prashanth-m) [inline mission/mission.proto]:
```
rpc SetCurrentMissionItemIndex(SetCurrentMissionItemIndexRequest) returns(SetCurrentMissionItemIndexResponse) {}
```

```
message SetCurrentMissionItemIndexRequest {
    int32 item_index = 1;
}
 message SetCurrentMissionItemIndexResponse {
     MissionResult mission_result = 1;
 }
```
How about this ?

File: mission/mission.proto (hunk 1 above)
Comment (shakthi-prashanth-m) [inline mission/mission.proto]: We need one for [mission_finished()](https://github.com/dronecore/DroneCore/blob/904ba588f554562672e79ea9dcc9449096219d47/plugins/mission/mission.h#L125).
Probably,
```
message IsMissionDoneRequest {}
message IsMissionDoneResponse {
     bool is_done = 1;
 }

rpc IsMissionDone(IsMissionDoneRequest) returns(IsMissionDoneResponse) {}
```

File: mission/mission.proto [hunk 4]
```diff
@@ -6,25 +6,63 @@ option java_package = "io.dronecore.mission";
...
+    rpc GetCurrentMissionItemId(GetCurrentMissionItemIdRequest) returns(GetCurrentMissionItemIdResponse) {}
```
Comment (shakthi-prashanth-m) [inline mission/mission.proto]:
```
message GetCurrentMissionItemIndexRequest {}
message GetCurrentMissionItemIndexResponse {
    int32 item_index = 1;
}

rpc GetCurrentMissionItemIndex(GetCurrentMissionItemIndexRequest) returns(GetCurrentMissionItemIndexResponse) {}
```

File: mission/mission.proto [hunk 5]
```diff
@@ -6,25 +6,63 @@ option java_package = "io.dronecore.mission";
...
+message PauseMissionResponse {
+    MissionResult mission_result = 1;
+}
+
+message SetCurrentMissionItemIdRequest {
+    int32 item_id = 1;
+}
+message SetCurrentMissionItemIdResponse {
+    MissionResult mission_result = 1;
+}
```
Comment (shakthi-prashanth-m) [inline mission/mission.proto]:
```
message SetCurrentMissionItemIndexRequest {
    int32 item_index = 1;
}
message SetCurrentMissionItemIndexResponse {
    MissionResult mission_result = 1;
}
```

File: mission/mission.proto (hunk 1 above)
Comment (shakthi-prashanth-m) [reply mission/mission.proto]: True, but we need a hook in DroneCore C++ interface.

Comment (JonasVautherin) [reply mission/mission.proto]: Good points. Nothing prevents us from extending the protocol later. So as soon as we have a "ClearMission" implementation on the DroneCore libs, we can add it to the protocol :-).

File: mission/mission.proto (hunk 3 above)
Comment (JonasVautherin) [reply mission/mission.proto]: Fine for me, I'll change it.

File: mission/mission.proto (hunk 2 above)
Comment (JonasVautherin) [reply mission/mission.proto]: Not sure about this one. But it's right that `MissionSize` might sound like "the distance covered by the mission" or something like this.

But I don't know if `MissionCount` is intuitive to me (but if it is for the others, I'm fine with it).

Or maybe `MissionItemsCount` or `MissionItemsNumber`?

I'd like to get more opinions on this one :-).

**Comment by shakthi-prashanth-m** [comment]
@JonasVautherin I've one question. If you see these [examples](https://github.com/grpc/grpc/blob/master/examples/protos/route_guide.proto), param names aren't extra-qualified. In case of message definition say,
```
message SetCurrentMissionItemIndexRequest {
    int32 index = 1;
}
```
is fine instead of saying `item_idex`, as it obviously refers to item,'s index.
What do you think ?

File: mission/mission.proto (hunk 2 above)
Comment (shakthi-prashanth-m) [reply mission/mission.proto]: Same as that of [MISSION_COUNT](http://mavlink.org/messages/common#MISSION_COUNT) in MAVLink.

Comment (shakthi-prashanth-m) [reply mission/mission.proto]: `MissionItemCount`  is also fine I think.

Comment (JonasVautherin) [reply mission/mission.proto]: If that's the MAVLink message, then I guess it makes sense to use it!

File: mission/mission.proto (hunk 1 above)
Comment (JonasVautherin) [reply mission/mission.proto]: `IsMissionFinished`, for consistency?

**Comment by JonasVautherin** [comment]
Fair point. I was not sure as it is a `*Request` (and in the route example, they really have structures), but I probably overthought this. I changed it for `index`, as you suggest.

And I applied the same for `GetMissionCountResponse` (`mission_count` -> `count`).

**Comment by shakthi-prashanth-m** [comment]
@JonasVautherin OK That makes sense. Thanks.

File: mission/mission.proto (hunk 1 above)
Comment (shakthi-prashanth-m) [reply mission/mission.proto]: Yeah, fine.

File: mission/mission.proto [hunk 6]
```diff
@@ -6,25 +6,69 @@ option java_package = "io.dronecore.mission";
...
+message IsMissionFinishedRequest {}
+message IsMissionFinishedResponse {
+    bool is_finished = 1;
 }
 
-message StartMissionRequest {
-    UUID uuid = 1;
+message SubscribeMissionProgressRequest {}
+message MissionProgressResponse {
+    int32 current_item_id = 1;
```
Comment (shakthi-prashanth-m) [inline mission/mission.proto]: `current_item_id` -> `current_item_index`

**Pull request merged**